SeqNo CAS linearizability

henningandersen opened the pull request:
Add a test that stresses concurrent writes using `ifSeqno/ifPrimaryTerm` to do CAS style updates. Use linearizability checker to verify linearizability using a simple sequential spec. Linearizability of successful CAS'es is guaranteed.

Changed linearizability checker to allow collecting history concurrently.

Changed unresponsive network simulation to wake up immediately when network disruption is cleared.

elasticmachine commented:
Pinging @elastic/es-distributed

henningandersen committed:
SeqNo CAS Linearizability check

Additional trace info when writing shard data.

henningandersen committed:
SeqNo CAS Linearizability check

Adjusted trace info.

henningandersen committed:
SeqNo CAS Linearizability check

Added linearizability checking for optimistic concurrency control using
ifSeqNo.

henningandersen commented:
@elasticmachine run elasticsearch-ci/1

henningandersen committed:
SeqNo CAS Linearizability check

Optimize memory usage of linearizability checker.

henningandersen committed:
SeqNo CAS Linearizability check

Checkstyle fix

henningandersen committed:
SeqNo CAS Linearizability check

Removed bad print statement.

henningandersen committed:
SeqNo CAS Linearizability check

Added linearizability check method that removes all timed out operations
and tests this first - optimistically hoping that this will succeed.
This helps avoid timing out tests for relatively trivial histories.

henningandersen committed:
SeqNo CAS Linearizability check

Checkstyle fixes.

henningandersen committed:
SeqNo CAS Linearizability check

Further spread seqno's around the best known value.

henningandersen commented on server/src/test/java/org/elasticsearch/cluster/coordination/LinearizabilityChecker.java:
```diff
@@ -0,0 +1,490 @@
+     * complete faster. Notice that the missingResponseGenerator here runs after partitioning, ie. input values will be the partitioned
+     * values
+     */
+    public boolean isLinearizableWithTimeoutOptimization(SequentialSpec spec, History history,
```
This is quite WIP and mainly included to allow us to discuss this optimization.

henningandersen replied:
Removed in 07f24c9602143d4f9cf1154c469c37d4423fb480

henningandersen commented on server/src/test/java/org/elasticsearch/cluster/coordination/LinearizabilityChecker.java:
```diff
@@ -0,0 +1,490 @@
+            return addInternal(internalizeStateMap.computeIfAbsent(state, k -> state), bitSet);
+        }
+
+        private boolean addInternal(Object state, FixedBitSet bitSet) {
```
We should add our own BitSet instead, since both BitSet and FixedBitSet has unnecessary memory overheads.

henningandersen committed:
SeqNo CAS Linearizability check

Minor fixes after rebase to master.

henningandersen committed:
SeqNo CAS Linearizability check

Fixed initial review comments. Serializable->Writeable, use +/- 1
primaryTerm rarely plus other minor changes.

henningandersen committed:
SeqNo CAS Linearizability check

Avoid negative seqNos.

henningandersen committed:
SeqNo CAS Linearizability check

Pick a new random disruption type per round. Added ability to visualize
a history.

henningandersen committed:
SeqNo CAS Linearizability check

Added additional trace info to IndexShard.index to get info on
success/failure of index requests.
Fixed an issue with the test: when a coordinating node retries, the CAS
fail on own write may not necessarily result in a new term, so have to
be more lenient in the version check. Improved comments to describe the
scenarios.

henningandersen commented on server/src/test/java/org/elasticsearch/versioning/ConcurrentSeqNoVersioningIT.java:
```diff
@@ -0,0 +1,686 @@
+                        final boolean futureSeqNo = pct < 10;
+                        Version version = partition.latestKnownVersion();
+                        if (futureSeqNo) {
+                            version = version.nextSeqNo(random.nextInt(4)+1);
```
Added in dc06ead10d4801037595901feeaba182b86a4d5a

henningandersen commented on server/src/test/java/org/elasticsearch/versioning/ConcurrentSeqNoVersioningIT.java:
```diff
@@ -0,0 +1,686 @@
+        // We can get CAS failures in following situations:
+        // 1. Real: a concurrent or previous write updated this.
+        // 2. Fail on our own write: the write to a replica was already done, but not responded to primary (or multiple replicas). And
+        // replica is promoted to primary, causing the write to fail on its own update.
```
Fixed in dc06ead10d4801037595901feeaba182b86a4d5a

henningandersen commented on server/src/test/java/org/elasticsearch/versioning/ConcurrentSeqNoVersioningIT.java:
```diff
@@ -0,0 +1,686 @@
+                    }
+
+                } catch (InterruptedException | BrokenBarrierException | TimeoutException e) {
+                    // check to stop
```
Added in dc06ead10d4801037595901feeaba182b86a4d5a (broken barrier is ok though).

henningandersen commented on server/src/test/java/org/elasticsearch/versioning/ConcurrentSeqNoVersioningIT.java:
```diff
@@ -0,0 +1,686 @@
+                return;
+            for (;;) {
+                Version current = this.current.get();
+                if (version.lteq(current) || this.current.compareAndSet(current, version)) {
```
Settled on updateAndGet in dc06ead10d4801037595901feeaba182b86a4d5a

henningandersen commented on server/src/test/java/org/elasticsearch/cluster/coordination/LinearizabilityChecker.java:
```diff
@@ -112,13 +129,18 @@
-    public static class History {
-        private final List<Event> events;
-        private int nextId;
+    public static class History implements Serializable {
```
Fixed in dc06ead10d4801037595901feeaba182b86a4d5a

henningandersen commented on server/src/test/java/org/elasticsearch/cluster/coordination/LinearizabilityChecker.java:
```diff
@@ -170,15 +204,20 @@ public void complete(Function<Object, Object> missingResponseGenerator) {
             final History history = new History();
             history.events.addAll(events);
-            history.nextId = nextId;
+            history.nextId = new AtomicInteger(nextId.get());
```
Fixed in dc06ead10d4801037595901feeaba182b86a4d5a

henningandersen commented on server/src/test/java/org/elasticsearch/versioning/ConcurrentSeqNoVersioningIT.java:
```diff
@@ -0,0 +1,686 @@
+    /**
+     * Our version, which is primaryTerm,seqNo.
+     */
+    private static final class Version implements Serializable {
```
Fixed in dc06ead10d4801037595901feeaba182b86a4d5a

henningandersen commented on server/src/test/java/org/elasticsearch/versioning/ConcurrentSeqNoVersioningIT.java:
```diff
@@ -0,0 +1,686 @@
+        }
+
+        private Version parseException(String message) {
+            // ugly, but having these observed versions available improves the linearizability checking and ensures progress.
```
Improved the comment in dc06ead10d4801037595901feeaba182b86a4d5a .

Using the version from the CAS failure exception ensures progress, since otherwise we would never be able to progress after just one CAS fail on own write.

henningandersen committed:
SeqNo CAS Linearizability check

Switched to using compareTo over gt.

henningandersen commented on server/src/test/java/org/elasticsearch/versioning/ConcurrentSeqNoVersioningIT.java:
```diff
@@ -0,0 +1,686 @@
+            return Objects.hash(primaryTerm, seqNo);
+        }
+
+        public boolean lteq(Version other) {
```
Fixed in dc06ead10d4801037595901feeaba182b86a4d5a and 096a91c52fc6469cc6c5dd09d3ddf8e3b08ce442

henningandersen committed:
SeqNo CAS Linearizability check

Added an additional simpler sequential spec that is easier to reason
about correctness of (but less precise).

henningandersen committed:
SeqNo CAS Linearizability check

Randomly use up to 3 replicas and 6 nodes. Be more specific about
timeout and try to eagerly connect after disruption to bring down
healing time.

henningandersen committed:
Merge remote-tracking branch 'origin/master' into seqno_cas_linearizability

henningandersen committed:
SeqNo CAS Linearizability check

Disabled delayed allocation in test (unnecessary).

henningandersen committed:
SeqNo CAS Linearizability check

Checkstyle fix

henningandersen committed:
SeqNo CAS Linearizability check

Improve runtime of ConcurrentSeqNoVersioningIT by resuming network
immediately when clearing network disruptions.

henningandersen committed:
Merge remote-tracking branch 'origin/master' into seqno_cas_linearizability

henningandersen committed:
SeqNo CAS Linearizability check

Avoid threads doing more work after they should stop.

henningandersen committed:
SeqNo CAS Linearizability check

Fix checkstyle issue

henningandersen committed:
SeqNo CAS Linearizability check

Reenable linearizability checkers timeout optimization for
CoordinatorTests.

henningandersen committed:
SeqNo CAS Linearizability check

Fix NPE in network disruptions.

henningandersen committed:
Merge remote-tracking branch 'origin/master' into seqno_cas_linearizability

henningandersen committed:
SeqNo CAS Linearizability check

Added missing equals/hashCode on SimpleState.

ywelsch commented on server/src/test/java/org/elasticsearch/versioning/ConcurrentSeqNoVersioningIT.java:
```diff
@@ -0,0 +1,905 @@
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        } catch (ClassCastException e) {
+            return "Unable to produce base64 serialized version of history: " + history +"\n  Message: " + e.getMessage();
```
should we fail in this case? This is unexpected, no?

Should we always check that serialization is not broken by always serializing histories, even if they have been linearizable?

henningandersen replied:
Fixed in 2f576d99d0d14bf375edc2158ce0a62210dfbe5e

ywelsch reviewed (changes requested):
Thank you for the work on this. I have left a lot of nits, just so that we can fully align on code style. I've also added a few suggestions for naming / describing states. The sequential specs make sense to me, might be good however to also add a few tests like was done for the other specs (see e.g. `testLockConsistent`, `testRegisterConsistent`, and `testRegisterSpecConsistency`). Let's also remove the WIP label in the PR title and update the PR description with a more detailed description of what's done in this PR.

henningandersen committed:
SeqNo CAS Linearizability check

Removed timeout optimization since we could risk seeing cases where it
would increase check time significantly. Also, maybe there are better
solutions to the problem, like hinting that a failure at a specific time
most likely means that state is not touched later.

henningandersen committed:
SeqNo CAS Linearizability check

Make visualize() static.

henningandersen committed:
SeqNo CAS Linearizability check

Rely on throwing exception out of main instead of handling it.

henningandersen committed:
SeqNo CAS Linearizability check

ie. -> i.e.,

henningandersen committed:
SeqNo CAS Linearizability check

Clearer description of the contents of VersionConflictEngineException.

henningandersen committed:
SeqNo CAS Linearizability check

Space nits.

henningandersen committed:
SeqNo CAS Linearizability check

Added missing braces.

henningandersen committed:
SeqNo CAS Linearizability check

Better logging of request failures.

henningandersen committed:
SeqNo CAS Linearizability check

lastKnownVersion field name.

henningandersen committed:
SeqNo CAS Linearizability check

Use scaledRandomInt.

henningandersen commented on server/src/test/java/org/elasticsearch/versioning/ConcurrentSeqNoVersioningIT.java:
```diff
@@ -0,0 +1,905 @@
+                } catch (TimeoutException e) {
+                    roundBarrier.reset();
+                }
+                internalCluster().clearDisruptionScheme(false);
```
I chose to only ensureFullyConnected over also waiting for healthy since that part seems unnecessary here. First of all, it waits for healthy before doing `ensureFullyConnected`, so I guess that could delay test a bit. Second, I think the wait here is unnecessary and I prefer not to have it since the test should continue to run regardless of how many nodes master thinks are in the cluster and whethere there are any relocating shards.
In summary: next round could start earlier without the wait, which seems like it could provoke more failures.

ywelsch replied:
ok

henningandersen committed:
SeqNo CAS Linearizability check

Favor slightly more readable code over slightly less randomization.

henningandersen committed:
SeqNo CAS Linearizability check

Reorder classes in more logical order.

henningandersen committed:
SeqNo CAS Linearizability check

Reuse SuccessState object when unmodified.

henningandersen committed:
SeqNo CAS Linearizability check

Reuse CASFailOwnWrite object when unmodified.

henningandersen committed:
SeqNo CAS Linearizability check

Do not swallow serialization exceptions and check that all histories
are serializable.

henningandersen committed:
SeqNo CAS Linearizability check

Renamed CASFailOwnWriteState to BoundedUncertaintyState and converted
State into interface to not share state and thus be specific about
fields in each state.

henningandersen committed:
SeqNo CAS Linearizability check

Fixed post test assertion on translog/lucene index consistency to not
run during recovery.

henningandersen committed:
SeqNo CAS Linearizability check

Brush up javadoc on BoundedUncertaintyState.

henningandersen committed:
SeqNo CAS Linearizability check

Improved explanation of expected CAS behaviour.

henningandersen committed:
SeqNo CAS Linearizability check

Javadoc checkstyle fix.

henningandersen commented on server/src/test/java/org/elasticsearch/cluster/coordination/LinearizabilityChecker.java:
```diff
@@ -267,6 +311,80 @@ public boolean isLinearizable(SequentialSpec spec, History history) {
+    /**
+     * Return a visual representation of the history
+     */
+    public String visualize(SequentialSpec spec, History history, Function<Object, Object> missingResponseGenerator) {
```
Fixed in ac73ca811380fffbf8b84af4fdf5099c2ece8e2c

henningandersen commented on server/src/test/java/org/elasticsearch/versioning/ConcurrentSeqNoVersioningIT.java:
```diff
@@ -0,0 +1,905 @@
+        } else {
+            try {
+                runLinearizabilityChecker(new FileInputStream(args[0]), Long.parseLong(args[1]), Long.parseLong(args[2]));
+            } catch (Exception e) {
```
Fixed in 9466acfef947369f9eb5dbe6090ec5b8e8ae34c1

henningandersen commented on server/src/test/java/org/elasticsearch/versioning/ConcurrentSeqNoVersioningIT.java:
```diff
@@ -0,0 +1,905 @@
+    // Wait up to 1 minute (+10s in thread to ensure it does not time out) for threads to complete previous round before initiating next
+    // round.
+    // Following issues are accepted for now:
+    // 1. Under certain circumstances (network partitions and other failures) we can end up giving a false negative response, ie. report
```
Sorry, bad habit, fixed in 6cdc8f2892f7f2705aaeef433118b5409f4f1ea9

henningandersen commented on server/src/test/java/org/elasticsearch/versioning/ConcurrentSeqNoVersioningIT.java:
```diff
@@ -0,0 +1,905 @@
+    // disruption.
+    // 4. Likewise, the CAS check can be done against a dirty write and thus fail even if it ought to succeed.
+    // 5. A CAS check can be done against a stale primary and fail due to that.
+    // 6. A CAS failure does not give back the current seqno/primary-term to use for the next request (and if it did, it could be wrong
```
Fixed in 8f08e5f0f03c2acf3f4c43eb3a2ac4d0fcef5ef2 but restructured comments in: 26ea11640ca4416f11a38412712f12d5b70339d1

henningandersen commented on server/src/test/java/org/elasticsearch/versioning/ConcurrentSeqNoVersioningIT.java:
```diff
@@ -0,0 +1,905 @@
+    // multiple threads doing CAS updates.
+    // Wait up to 1 minute (+10s in thread to ensure it does not time out) for threads to complete previous round before initiating next
+    // round.
+    // Following issues are accepted for now:
```
Fixed in 26ea11640ca4416f11a38412712f12d5b70339d1

henningandersen commented on server/src/test/java/org/elasticsearch/versioning/ConcurrentSeqNoVersioningIT.java:
```diff
@@ -0,0 +1,905 @@
+        ensureGreen();
+
+
+        int numberOfKeys = randomIntBetween(1,10);
```
Fixed in 5f2465e934e6bbac1a0ba1765e6843b2331b9ad4

henningandersen commented on server/src/test/java/org/elasticsearch/versioning/ConcurrentSeqNoVersioningIT.java:
```diff
@@ -0,0 +1,905 @@
+
+        List<CASUpdateThread> threads =
+            IntStream.range(0, threadCount)
+                .mapToObj(i -> new CASUpdateThread(i, roundBarrier, partitions, disruptTimeSeconds+1))
```
Fixed in 5f2465e934e6bbac1a0ba1765e6843b2331b9ad4

henningandersen commented on server/src/test/java/org/elasticsearch/versioning/ConcurrentSeqNoVersioningIT.java:
```diff
@@ -0,0 +1,905 @@
+                try {
+                    roundBarrier.await(70, TimeUnit.SECONDS);
+
+                    int numberOfUpdates = randomIntBetween(3*partitions.size(), 13*partitions.size());
```
Fixed in 5f2465e934e6bbac1a0ba1765e6843b2331b9ad4

henningandersen commented on server/src/test/java/org/elasticsearch/versioning/ConcurrentSeqNoVersioningIT.java:
```diff
@@ -0,0 +1,905 @@
+            this.lastFailed = lastFailed;
+        }
+
+
```
Fixed in 5f2465e934e6bbac1a0ba1765e6843b2331b9ad4

henningandersen commented on server/src/test/java/org/elasticsearch/versioning/ConcurrentSeqNoVersioningIT.java:
```diff
@@ -0,0 +1,905 @@
+        }
+
+        public Version previousTerm() {
+            return new Version(primaryTerm-1, seqNo);
```
Fixed in 5f2465e934e6bbac1a0ba1765e6843b2331b9ad4

henningandersen commented on server/src/test/java/org/elasticsearch/versioning/ConcurrentSeqNoVersioningIT.java:
```diff
@@ -0,0 +1,905 @@
+        }
+
+        public Version nextTerm() {
+            return new Version(primaryTerm+1, seqNo);
```
Fixed in 5f2465e934e6bbac1a0ba1765e6843b2331b9ad4

henningandersen commented on server/src/test/java/org/elasticsearch/versioning/ConcurrentSeqNoVersioningIT.java:
```diff
@@ -0,0 +1,905 @@
+                            // if we used a future seqNo, we cannot know if it will overwrite a future update when failing with
+                            // unknown error, so have to assume a timeout instead wrt. linearizability.
+                            if (futureSeqNo == false && futureTerm == false)
+                                historyResponse.accept(new FailureHistoryOutput());
```
Fixed in 113356fe4fe363dd978244eb849ee5df8be6ae9c

henningandersen commented on server/src/test/java/org/elasticsearch/versioning/ConcurrentSeqNoVersioningIT.java:
```diff
@@ -0,0 +1,905 @@
+
+        @Override
+        Optional<State> casSuccess(Version inputVersion, Version outputVersion) {
+            if (safeVersion.compareTo(inputVersion) > 0 || safeVersion.compareTo(outputVersion) > 0)
```
Fixed in 113356fe4fe363dd978244eb849ee5df8be6ae9c

henningandersen commented on server/src/test/java/org/elasticsearch/versioning/ConcurrentSeqNoVersioningIT.java:
```diff
@@ -0,0 +1,905 @@
+                            // unknown error, so have to assume a timeout instead wrt. linearizability.
+                            if (futureSeqNo == false && futureTerm == false)
+                                historyResponse.accept(new FailureHistoryOutput());
+                            logger.info("Received failure: " + e.getMessage(), e);
```
Fixed in 7fcf7edbb129e8fe88906b3a7b20e6804c177daa

henningandersen commented on server/src/test/java/org/elasticsearch/versioning/ConcurrentSeqNoVersioningIT.java:
```diff
@@ -0,0 +1,905 @@
+
+    private class Partition {
+        private final String id;
+        private final AtomicVersion atomicVersion;
```
Fixed in 81540fc75af85cfae1fdb68092b9602a689eaf78

henningandersen commented on server/src/test/java/org/elasticsearch/versioning/ConcurrentSeqNoVersioningIT.java:
```diff
@@ -0,0 +1,905 @@
+    // due to a dirty write anyway).
+    public void testSeqNoCASLinearizability() {
+        // 1-8 seconds, bias towards shorter.
+        final int disruptTimeSeconds = 1 << randomInt(3);
```
Fixed in 58e478af0a8076ee2ca10a51ea6149c1e239c24a

henningandersen commented on server/src/test/java/org/elasticsearch/versioning/ConcurrentSeqNoVersioningIT.java:
```diff
@@ -0,0 +1,905 @@
+                try {
+                    roundBarrier.await(70, TimeUnit.SECONDS);
+
+                    int numberOfUpdates = randomIntBetween(3*partitions.size(), 13*partitions.size());
```
Fixed in 9a03fc1c520091e4fe5256a404b301db08c62e73

henningandersen commented on server/src/test/java/org/elasticsearch/versioning/ConcurrentSeqNoVersioningIT.java:
```diff
@@ -0,0 +1,905 @@
+
+    }
+
+    private static class CASSequentialSpec implements LinearizabilityChecker.SequentialSpec {
```
Fixed in bd4774d9aa489f02cc16d7c4f5cfb47d0070e7c1 , did a few more reorderings of inner classes.

henningandersen commented on server/src/test/java/org/elasticsearch/versioning/ConcurrentSeqNoVersioningIT.java:
```diff
@@ -0,0 +1,905 @@
+                // ignored. It cannot have made surviving interim writes, since if so, the final primaryTerm should be greater than or
+                // equal to safeVersion.primaryTerm
+                // So the last successful write is still the right representation for the state.
+                return Optional.of(new SuccessState(safeVersion));
```
Yes, of course, fixed in 1b55ba993df53b19aaee15fe84b8890fbca84bed

henningandersen commented on server/src/test/java/org/elasticsearch/versioning/ConcurrentSeqNoVersioningIT.java:
```diff
@@ -0,0 +1,905 @@
+            // We use the max cas fail version, which is correct due to the checking in casSuccess.
+            // for 4, we ignore the stale read (does not affect state).
+
+            Version newCASFailVersion = outputVersion.compareTo(casFailVersion) > 0 ? outputVersion : casFailVersion;
```
Fixed in bb014e7ad51791943d101621a2feb66f07f6a975

henningandersen commented on server/src/test/java/org/elasticsearch/versioning/ConcurrentSeqNoVersioningIT.java:
```diff
@@ -0,0 +1,905 @@
+     *
+     * Additionally following CAS failures on top of this state stays in this state, see #casFail method.
+     */
+    private static class CASFailOwnWriteState extends State {
```
I agree, thanks for the suggestion. Fixed in e5f9d709e49e3da7a36bf981ae4cc915791e132c

henningandersen committed:
SeqNo CAS Linearizability check

Added test case for complex sequential spec.

henningandersen committed:
SeqNo CAS Linearizability check

Added test case for simple sequential spec.

henningandersen commented:
Thanks @ywelsch , I believe I have addressed all points now, so please have another look.

henningandersen committed:
SeqNo CAS Linearizability check

Strengthened complex sequential spec a bit by not allowing illegal
stale reads.

henningandersen committed:
Merge remote-tracking branch 'origin/master' into seqno_cas_linearizability

henningandersen committed:
SeqNo CAS Linearizability check

Removed complex sequential spec, since the stronger guarantees it claims
can be violated. Both CAS failure and other failures can eventually
succeed any time after reciving the response - until a successful CAS
has responded with a higher version.

Fixed bug around detecting if we can early terminate a failed operation.

Updated comments to reflect this new knowledge.

henningandersen committed:
SeqNo CAS Linearizability check

Removed newly added unused method.

henningandersen committed:
SeqNo CAS Linearizability check

Whitespace improvements.

ywelsch reviewed (approved):
LGTM

henningandersen committed:
SeqNo CAS Linearizability check

Removed ClassCastException catch, leftover from using serializable.

henningandersen commented on server/src/test/java/org/elasticsearch/versioning/ConcurrentSeqNoVersioningIT.java:
```diff
@@ -0,0 +1,1062 @@
+            }
+            output.close();
+            return Base64.getEncoder().encodeToString(BytesReference.toBytes(output.bytes()));
+        } catch (IOException | ClassCastException e) {
```
Ooops, thanks, fixed in 08286ebb6e44f02ce80cb3ff2b84f86d9d69199f

henningandersen commented:
Failed test seems unrelated, reported it here: https://github.com/elastic/elasticsearch/issues/41549

@elasticmachine run elasticsearch-ci/1

henningandersen commented:
Another unrelated test failure, reported here: #41566

@elasticmachine run elasticsearch-ci/1

henningandersen committed:
Merge remote-tracking branch 'origin/master' into seqno_cas_linearizability

henningandersen commented:
@elasticmachine run elasticsearch-ci/1

henningandersen commented:
Unrelated failure, report here: #41256
@elasticmachine run elasticsearch-ci/1
